src/shared/metrics.c

```diff
@@ -0,0 +1,242 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "json-util.h"
+#include "log.h"
+#include "metrics.h"
+#include "string-table.h"
+#include "strv.h"
+#include "varlink-io.systemd.Metrics.h"
+#include "varlink-serialize.h"
+#include "varlink-util.h"
+
+static void metric_family_context_done(MetricFamilyContext *ctx) {
+        assert(ctx);
+
+        sd_json_variant_unref(ctx->previous);
+}
+
+int metrics_setup_varlink_server(
+                sd_varlink_server **server, /* in and out param */
+                sd_varlink_server_flags_t flags,
+                sd_event *event,
+                sd_varlink_method_t vl_method_list_cb,
+                sd_varlink_method_t vl_method_describe_cb,
+                void *userdata) {
+
+        _cleanup_(sd_varlink_server_unrefp) sd_varlink_server *s = NULL;
+        int r;
+
+        assert(server);
+        assert(event);
+
+        if (*server)
+                return 0;
+
+        r = varlink_server_new(&s, flags, userdata);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to allocate varlink metrics server object: %m");
+
+        r = sd_varlink_server_add_interface(s, &vl_interface_io_systemd_Metrics);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to add varlink metrics interface to varlink server: %m");
+
+        r = sd_varlink_server_bind_method_many(
+                s,
+                "io.systemd.Metrics.List", vl_method_list_cb,
+                "io.systemd.Metrics.Describe", vl_method_describe_cb);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to register varlink metrics methods: %m");
+
+        r = sd_varlink_server_attach_event(s, event, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink metrics connection to event loop: %m");
+
```

Comment:
please add sd_varlink_server_set_description() to set a name here, for debugging purposes